### PR TITLE
Adds PII/Profanity failure exception cases on AI assessment

### PR DESCRIPTION
**Pull request by wilkie**
Adds checks using `ShareFiltering` to ensure nothing is sent to AI assessment that fails basic PII checks (and profanity checks).

* This refactors the `ShareFiltering` module to allow for exceptions since exceptions trigger the failure case for the job, and it makes sense to have the PII check throw one.
* Refactors `ShareFiltering` such that PII failures happen first before a profanity check since it would send some PII to the third-party profanity service, which seemed counter-intuitive. 
* Refactors to separate PII and profanity so you can elect to query either of them, alone. This was due to us initially only wanting a PII check, but it seemed useful to keep since PII checks are arguably more important and, in our code as it stands, much cheaper to check.

`RubricAiEvaluation` model

* Since this adds complexity to statuses (by adding several different failure cases with reasons), the status field has been expanded to have failures be any value over 1000.
* The `RubricAiEvaluation` model contains helper functions to negotiate which statuses are success and failure.

`EvaluateRubricJob` job

* Adds `rescue_from` hooks to the `EvaluateRubricJob` job for each type of share violation.

## Links

- jira ticket: [AITT-189](https://codedotorg.atlassian.net/browse/AITT-189)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/test/controllers/api/v1/ml_models_controller_test.rb [hunk 1]
```diff
@@ -5,8 +5,6 @@ def stub_firehose
     FirehoseClient.instance.stubs(:put_record).with do |stream, args|
       @firehose_record = args
       @firehose_stream = stream
-      puts args
-      puts stream
```
Comment (wilkie) [inline dashboard/test/controllers/api/v1/ml_models_controller_test.rb:9]: This is just noise in the test log. I ran into it while testing this. This module also tests its integration with the PII/Profanity detection.

File: dashboard/app/jobs/evaluate_rubric_job.rb [hunk 2]
```diff
@@ -104,6 +138,12 @@ def perform(user_id:, requester_id:, script_level_id:, rubric_ai_evaluation_id:
     channel_id = get_channel_id(user, script_level)
     code, project_version = read_user_code(channel_id)
 
+    # Check for PII / sharing failures
+    # Get the 2-character language code from the user's preferred locale
+    locale = (user.locale || 'en')[0...2]
+    share_failure = ShareFiltering.find_share_failure(code, locale, exceptions: true)
+    raise "ShareFailure #{share_failure.type}" if share_failure
```
Comment (wilkie) [inline dashboard/app/jobs/evaluate_rubric_job.rb]: This doesn't need to be here since I've added exceptions directly to `ShareFiltering`. I'll probably remove it... it would cause a general failure... but it isn't possible.

File: lib/cdo/shared_constants.rb [hunk 3]
```diff
@@ -615,8 +615,12 @@ module SharedConstants
     RUNNING: 1,
     # Succeeded
     SUCCESS: 2,
-    # General failure
-    FAILURE: 3,
+    # General failure (any anything larger)
```
Comment (wilkie) [inline lib/cdo/shared_constants.rb]: Typo.

File: dashboard/app/jobs/evaluate_rubric_job.rb [hunk 4]
```diff
@@ -86,6 +86,40 @@ class EvaluateRubricJob < ApplicationJob
...
+      puts "EvaluateRubricJob Filter Error: #{exception.full_message} Type: #{exception.share_failure.type}"
+    end
+
+    # Record the failure, if we can
+    begin
+      rubric_ai_evaluation = pass_in_or_create_rubric_ai_evaluation(self)
+      rubric_ai_evaluation.status = SharedConstants::RUBRIC_AI_EVALUATION_STATUS[:PROFANITY_VIOLATION]
+      rubric_ai_evaluation.save!
+    rescue StandardError
+      # Ignore cascading errors when the rubric record does not exist
```
Comment (davidsbailey) [inline dashboard/app/jobs/evaluate_rubric_job.rb:100]: for this error, perhaps letting it through to honeybadger would be more appropriate, since something seems broken if the rubric_ai_evaluation does not exist. here and below

File: dashboard/app/jobs/evaluate_rubric_job.rb [hunk 5]
```diff
@@ -86,6 +86,40 @@ class EvaluateRubricJob < ApplicationJob
...
+    begin
+      rubric_ai_evaluation = pass_in_or_create_rubric_ai_evaluation(self)
+      rubric_ai_evaluation.status = SharedConstants::RUBRIC_AI_EVALUATION_STATUS[:PII_VIOLATION]
+      rubric_ai_evaluation.save!
+    rescue StandardError
+      # Ignore cascading errors when the rubric record does not exist
+    end
+
+    # We gracefully just fail, here, and we do not file this exception
+  end
```
Comment (davidsbailey) [inline dashboard/app/jobs/evaluate_rubric_job.rb:121]: my guess is that the rescue_from blocks get executed in order. are you seeing both "EvaluateRubricJob Error" and "EvaluateRubricJob Filter Error" get printed now? If so, and if it doesn't break anything, maybe it would make more sense (less debug noise) to have the new rescue_from calls happen before the original one above?

File: dashboard/app/models/rubric_ai_evaluation.rb [hunk 6]
```diff
@@ -28,4 +28,32 @@ class RubricAiEvaluation < ApplicationRecord
...
+
+  def pii_failure?
+    status == SharedConstants::RUBRIC_AI_EVALUATION_STATUS[:PII_VIOLATION]
+  end
+
+  def profanity_failure?
+    status == SharedConstants::RUBRIC_AI_EVALUATION_STATUS[:PROFANITY_VIOLATION]
+  end
+
+  def share_failure?
```
Comment (davidsbailey) [inline dashboard/app/models/rubric_ai_evaluation.rb]: what do you think about naming this `share_filtering_failure?` (or something else involving "filter") to be a bit more descriptive, since what we're doing has more to do with filtering than sharing?

**Review by davidsbailey — APPROVED** [review]
looks good, nice code reuse, great tests!

File: dashboard/app/jobs/evaluate_rubric_job.rb (hunk 5 above)
Comment (wilkie) [reply dashboard/app/jobs/evaluate_rubric_job.rb:121]: Oh that's something I should confirm. I think I misread some documentation and I did intend to test that. It is intended that one happen and not hit the general case.

Comment (wilkie) [reply dashboard/app/jobs/evaluate_rubric_job.rb:121]: You can see my confusion: https://edgeapi.rubyonrails.org/classes/ActiveSupport/Rescuable/ClassMethods.html

> Handlers are inherited. They are searched from right to left, ***from bottom to top***, and up the hierarchy.

I assume they are ALL executed if they match and it isn't just finding the FIRST one. But I'm not sure why it would do the general error first if it goes bottom to top.

File: dashboard/app/jobs/evaluate_rubric_job.rb (hunk 4 above)
Comment (wilkie) [reply dashboard/app/jobs/evaluate_rubric_job.rb:100]: It would shadow the original exception... and if an error occurs before the rubric ai evaluation was made, then the reason the record wouldn't exist is because of the original error. my thought was that the originating error was the important one to log.

An RAE record doesn't exist when it cannot find the ScriptLevel or Rubric or User and if any channel id errors happen... because the rubric evaluation record requires the project_id and rubric_id. These things shouldn't happen in practice because the controller action that is scheduling the jobs should also have to look these up to authorize the action.

Comment (wilkie) [reply dashboard/app/jobs/evaluate_rubric_job.rb:100]: Only the final exception makes it to honey badger

**Comment by abdurakhmonov049** [comment]
wonderful...

File: dashboard/app/models/rubric_ai_evaluation.rb (hunk 6 above)
Comment (wilkie) [reply dashboard/app/models/rubric_ai_evaluation.rb]: done

File: lib/cdo/shared_constants.rb (hunk 3 above)
Comment (wilkie) [reply lib/cdo/shared_constants.rb]: fixed.

File: dashboard/app/jobs/evaluate_rubric_job.rb (hunk 2 above)
Comment (wilkie) [reply dashboard/app/jobs/evaluate_rubric_job.rb]: I'll remove it along with the assignment.

Comment (wilkie) [reply dashboard/app/jobs/evaluate_rubric_job.rb]: Removed. I'm using exceptions for a reason.

File: dashboard/app/jobs/evaluate_rubric_job.rb (hunk 5 above)
Comment (davidsbailey) [reply dashboard/app/jobs/evaluate_rubric_job.rb:121]: > You can see my confusion: https://edgeapi.rubyonrails.org/classes/ActiveSupport/Rescuable/ClassMethods.html
> 
> > Handlers are inherited. They are searched from right to left, _**from bottom to top**_, and up the hierarchy.
> 
> I assume they are ALL executed if they match and it isn't just finding the FIRST one. But I'm not sure why it would do the general error first if it goes bottom to top.

Yes, it is rather confusing and I can't say I totally understand what the docs are saying should happen here. the next line from the docs are:

> The handler of the first class for which exception.is_a?(klass) holds true is the one invoked, if any.

which suggests that only one of your `rescue_from` blocks will get called (in your case the later, more specific ones in case of PIIFilterException). then it says:

> Exceptions raised inside exception handlers are not propagated up.

I am assuming this means that they are not propagated up (within the file) to other rescue_from handlers, since we already know from the existing code that reraised exceptions make it "out" of the rescue_from block and into honeybadger.

File: dashboard/app/jobs/evaluate_rubric_job.rb (hunk 4 above)
Comment (davidsbailey) [reply dashboard/app/jobs/evaluate_rubric_job.rb:100]: hmm are you saying the original exception (ProfanityFilterException) will automatically be reraised by the rescue_from block? I would bet that this is not the case because it seems dissimilar to the concept of rescuing in ruby.

> These things shouldn't happen in practice because the controller action that is scheduling the jobs should also have to look these up to authorize the action.

If something should not happen in practice, then we should surface exception when it does happen rather than squashing it, since it means an error on our part in either programming or design.

File: dashboard/app/jobs/evaluate_rubric_job.rb (hunk 5 above)
Comment (davidsbailey) [reply dashboard/app/jobs/evaluate_rubric_job.rb:121]: in any case I'm happy to retract my comment on changing the order based on the possibility of double debug statements, but I do think it is worth confirming that the new handlers are being called and taking effect, and the right things are being logged to honey badger.

Comment (wilkie) [reply dashboard/app/jobs/evaluate_rubric_job.rb:121]: Ok. Tested this. Learning a lot.

I guarded the setting of the PII_PROFANITY status to happen only if the status is not already FAILURE. If the StandardError happens first, the PII_PROFANITY status does not get set, so the tests fail. For good measure, guarded the normal error path so it doesn't set FAILURE status if it is set to a PII error status.

With `StandardError` on top, tests still pass... **confirming the order is bottom to top**.

Putting the `StandardError` lambda on the /bottom/ causes the two PII failure tests to fail because the exceptions bubble out unexpectedly. That suggests that the `StandardError` happens (bottom-most) and it re-raises and it doesn't perform any of the other `rescue_from`'s.

To sum up:

Defined in this order with print statements:

```
rescue_from(ProfanityFilterException)
rescue_from(PIIFilterException)
rescue_from(StandardError)
```

Produced:

```
StandardError
StandardError
StandardError
```

So they match bottom to top and once they match, they do not continue to match. So you want your most specific cases on the bottom and catch-alls at the top of the module.

Comment (wilkie) [reply dashboard/app/jobs/evaluate_rubric_job.rb:121]: https://www.vzhima.com/2019/12/02/rails-multiple-rescue-from.html

**Review by wilkie — COMMENTED** [review]
Seems like it is good to go.

File: dashboard/app/jobs/evaluate_rubric_job.rb (hunk 5 above)
Comment (davidsbailey) [reply dashboard/app/jobs/evaluate_rubric_job.rb:121]: Wonderful! Thank you for getting to the bottom of this! 

> Produced:
> 
> ```
> StandardError
> StandardError
> StandardError
> ```

What does each `StandardError` represent here?
